Fix issues with quotes and character encodings

bolner opened the pull request:
- Introduce 2 parsing states: inside quote, inside REM block
- Suppose 'IBM Code Page 437' encoding by default

Fixes #1

bolner committed:
Fix for issue: https://github.com/danvk/gwbasic-decoder/issues/1

- Introduce 2 parsing states: inside quote, inside REM block
- Suppose 'IBM Code Page 437' encoding by default

danvk commented:
Hi @bolner, thanks for the fix! I'm glad you've gotten value out of this tool.

Would you mind:

1. Adding a test file with expected output that demonstrates the fix.
2. At least make the encoding a parameter an argument. You're right that it should be a flag, but then that means that there should be support for flags in general :)

bolner commented:
Hi @danvk, yes, I'll do that today or tomorrow. Later I also realized other issues: like the single quote is also an alias for the REM. Plus: one of the files I attached is broken (probably data corruption due to age :) I'll try to fix it with a hex editor on the weekend), but the program doesn't detect it, it only displays the 20% of the lines (without error message). And I have a bad feeling about some of the floating point conversions. I think sometimes it would need rounding at less digits, to show the same numbers as the original were. I could test that in an emulator.
Also it doesn't run on Python 3, only on 2. Should that stay this way, or can I change it to Python 3? :) (I think it wouldn't support 2 anymore)

danvk commented:
Feel free to make it Python 3. In fact, please make it Python 3!

bolner committed:
- Changed the project to Python 3
- Committed a "launch.json" for VS Code, which can be used for debugging
- Added the encoding as a new command line parameter
- It outputs now usage info if called without parameters
- Added static typing
- Renamed the "bytes" variable to "data", so that variable won't have the same name as an existing type, which can be confusing.
- The data and the encoding are now passed to the objects in their constructors.
- Got rid of the static "factory methods".
- Rewrote the float conversion to avoid losing accuracy and implemented the double conversion.
- Renamed gwbasicline to gwBasicLine
- Moved the token parsing functions into gwBasicLine
- Renamed gwbasic to gwBasic
- It now can parse the single quote alias of the REM command (stored as 3 bytes in reality)
- The floats in GAME.BAS are now properly displayed. Updated the readme.

bolner commented:
@danvk Hi, I'm not ready yet, so I've created a dev branch: https://github.com/bolner/gwbasic-decoder/tree/dev

I've implemented the doubles, and fixed some float issues (The 0.8 in GAME.BAS is okay now), but rounding and formatting is still not okay everywhere. I've added a `TYPEDEMO.BAS` which demonstrates many types. Its output should be as on the screenshot I attached (It's from a real GW-Basic):
![type_demo](https://user-images.githubusercontent.com/2199459/64825486-babd1580-d5bd-11e9-971d-389bf1c30015.png)

So the hex / octal types are neither showing up properly.
I think I can finish it on the weekend :)

List of all changes:
- Changed the project to Python 3
- Committed a "launch.json" for VS Code, which can be used for debugging
- Added the encoding as a new command line parameter
- It outputs now usage info if called without parameters
- Added static typing
- Renamed the "bytes" variable to "data", so that variable won't have the same name as an existing type, which can be confusing.
- The data and the encoding are now passed to the objects in their constructors.
- Got rid of the static "factory methods".
- Rewrote the float conversion to avoid losing accuracy and implemented the double conversion.
- Renamed gwbasicline to gwBasicLine
- Moved the token parsing functions into gwBasicLine
- Renamed gwbasic to gwBasic
- It now can parse the single quote alias of the REM command (stored as 3 bytes in reality)
- The floats in GAME.BAS are now properly displayed. Updated the readme.

bolner committed:
Float/Double fixed!

bolner committed:
Fix octal/hex, add more sample files, extend README

bolner committed:
Merge branch 'dev'

bolner committed:
case problem

bolner commented:
@danvk Now everything works which was planned, but I've noticed another bug. Did you know that the `GAME.BAS` has 7 more lines? :) Just checked it in an emulator. This problem existed before my changes as well. I'll check it.

bolner committed:
Fix EOF issue, add boundary checks

bolner commented:
@danvk Fixed everything, and tested with many BAS files, other than the ones in README.md. No more TODOs left :) Please have a look when you have some free time. Everything is in master.

bolner committed:
README: fix GAME.BAS output

bolner committed:
Fix for single quote boundary check

bolner committed:
Change Shebang to Python 3

bolner committed:
More exact formatting +remove unnecessary exec flags

bolner committed:
remove unnecessary exec flags

danvk commented on convert.py:
```diff
@@ -1,13 +1,41 @@
-        program = gwbasic.FromBinary(data)
-        print program
+    # Check the number of parameters
+    if not (2 <= len(sys.argv) <= 3):
```
Suggest making this a flag and using Python's built-in [argparse][1], which supports optional arguments, flags and values which have to be one of a set of choices. This sort of argument processing has a way of getting very messy as the number of parameters inevitably increases.

[1]: https://docs.python.org/3/library/argparse.html

bolner replied:
Just tried it, looks good, will commit soon.

danvk commented on convert.py:
```diff
@@ -1,13 +1,41 @@
+            program.Parse()
+            print(program)
+
+    except Exception as ex:
```
Suggest removing this try / except — it seems almost exactly the same as the behavior without it?

bolner replied:
Hmm yes :) Except for the extra newlines around it, but that might not worth the complication. Will remove.

danvk commented on gwbasic.py:
```diff
@@ -1,144 +1,236 @@
-        return [self, pos]
+
+        self.pos += 1    # consume the null byte
+        return
```
this is a no-op

bolner replied:
You mean the "return"? Yes, didn't notice, will remove it.

danvk commented on gwbasic.py:
```diff
@@ -1,144 +1,236 @@
+                continue
+            elif code == 0x3a and not (insideQuotes or insideRem):
+                if len(self.data) - self.pos - 1 > 2:
+                    if self.data[self.pos + 1] == 0x8f and self.data[self.pos + 2] == 0xd9:
```
combine these into a single `if`

bolner replied:
Ah yes, I wasn't sure that Python supports this: https://en.wikipedia.org/wiki/Short-circuit_evaluation   Then we would also not need the 2 continues.

danvk commented on gwbasic.py:
```diff
@@ -1,144 +1,236 @@
-
-
-class gwbasicline:
+class gwBasicLine:
```
I realize you're inheriting this, but in Python you generally CamelCase class names, so `GWBasicLine` and `GWBasic`.

danvk reviewed (approved):
A few comments, but otherwise looks great. I'm going to add you as an owner of this repo since you've written more of the code than I have at this point!

bolner committed:
+argparse, class naming and minor changes

bolner commented:
@danvk Committed changes.

bolner committed:
minor change
